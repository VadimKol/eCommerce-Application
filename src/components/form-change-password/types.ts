export interface CustomerProfile {
  version: number;
  firstName: string;
  lastName: string;
  dateOfBirth: string;
  email: string;
  password: string;
  billingAddressIds: string[];
  shippingAddressIds: string[];
  defaultShippingAddressId: string;
  defaultBillingAddressId: string;
}

export interface FormChangePasswordProps {
  email: string;
  version: number;
  setPersonInfo: (personInfo: CustomerProfile) => void;
}
