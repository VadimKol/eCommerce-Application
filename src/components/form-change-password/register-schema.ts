import { z } from 'zod';

export const registerSchema = z.object({
  currentPassword: z
    .string()
    .min(8, 'Minimum 8 characters')
    .regex(/[A-Z]/, 'Password must contain at least one uppercase letter')
    .regex(/[a-z]/, 'Password must contain at least one lowercase letter')
    .regex(/[0-9]/, 'Password must contain at least one digit')
    .refine(
      (val) => val[0] !== ' ' && val[val.length - 1] !== ' ',
      'Password must not contain leading or trailing whitespace.',
    ),
  newPassword: z
    .string()
    .min(8, 'Minimum 8 characters')
    .regex(/[A-Z]/, 'Password must contain at least one uppercase letter')
    .regex(/[a-z]/, 'Password must contain at least one lowercase letter')
    .regex(/[0-9]/, 'Password must contain at least one digit')
    .refine(
      (val) => val[0] !== ' ' && val[val.length - 1] !== ' ',
      'Password must not contain leading or trailing whitespace.',
    ),
});

export type FormValues = z.infer<typeof registerSchema>;
