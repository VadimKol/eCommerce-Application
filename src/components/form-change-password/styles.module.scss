
.form {
  display: flex;
  flex-flow: column nowrap;
  justify-content: space-around;
  align-items: center;
  box-sizing: border-box;
  width: calc(100% - 40px);
  max-width: 600px;
  margin: 50px auto;
  padding: 0 50px 10px;
  font-size: 16px;
  color: var(--primary-color);
}

.inputWithError {
  padding-bottom: 10px;
}

.formInput {
  display: flex;
  flex-flow: column nowrap;
  justify-content: space-between;
  align-items: start;
  width: 100%;
}

.input {
  width: 100%;
  height: 30px;
  padding: 0 5px;
  font-family: var(--primary-font);
  color: var(--primary-color);
  border: 1px solid var(--primary-color);
  border-radius: 5px;
}

.errorMsg {
  width: 100%;
  margin: -20px 0 -10px;
  font-size: 14px;
  text-align: right;
  color: var(--error-color);
}

.changeInfo {
  width: 80px;
  height: 40px;
  margin-top: 30px;
  font-size: 1em;
  color: var(--text-color);
  background: var(--primary-color);
  border: 1px solid var(--primary-color);
  border-radius: 6px;
  cursor: pointer;
  transition: all 500ms;
}

.changeInfo:hover {
  box-shadow: -5px 5px 5px var(--secondary-color);
  transition: all 500ms;
  transform: translateY(-2px);
}

.invalid,
.input.invalid:focus {
  box-shadow: 0 0 7px var(--error-color);
}

.input:focus {
  border: 1px solid var(--primary-color);
  box-shadow: 0 0 7px var(--primary-color);
}

.disabled {
  background-color: var(--disabled-color);
  border-color: var(--disabled-color);
  pointer-events: none;
  user-select: none;
}

.valid {
  box-shadow: 0 0 7px var(--secondary-color);
}

.invalid {
  box-shadow: 0 0 7px var(--error-color);
}

.error {
  width: 100%;
  margin-bottom: 20px;
  color: var(--error-color);
}

.titleWithTooltip {
  display: flex;
  flex-flow: nowrap;
  gap: 20px;
  justify-content: center;
  align-items: center;
}
