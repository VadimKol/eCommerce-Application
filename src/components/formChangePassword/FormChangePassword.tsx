import type { ClientResponse, Customer, MyCustomerChangePassword } from '@commercetools/platform-sdk';
import { useState } from 'react';
import { toast } from 'react-toastify';

import { tokenCache } from '@/api/build-client';
import { changePassword, login } from '@/api/client-actions.ts';

import styles from './styles.module.scss';
import type { FormChangePasswordProps } from './types.ts';

export function FormChangePassword({ email, version }: FormChangePasswordProps): JSX.Element {
  const [currentPassword, setCurrentPassword] = useState<string>('');
  const [newPassword, setNewPassword] = useState<string>('');

  const changePasswordHandle = (): void => {
    const myCustomerChangePassword: MyCustomerChangePassword = {
      version,
      currentPassword,
      newPassword,
    };
    changePassword(myCustomerChangePassword)
      .then((response: ClientResponse<Customer>) => {
        if (response) {
          toast('Previous password entered correctly', { type: 'success' });
        }
      })
      .then(() => {
        const password = myCustomerChangePassword.newPassword;

        login({ email, password })
          .then(() => {
            localStorage.setItem('geek-shop-token', `${tokenCache.get().token}`);
            toast(`Password change was successful`, { type: 'success' });
          })
          .catch(() => {
            toast('An error occurred when changing your password, please try again', { type: 'error' });
          });
      })
      .catch((err: Error) => {
        toast(`Password error ${err.message}`, { type: 'error' });
      });
  };

  return (
    <div className={styles.blockPassword}>
      <h2>Change password</h2>
      <div>
        To change your password, enter your current password and the new one. Once the current password has been
        successfully verified, it will be replaced with a new password.
      </div>
      <label htmlFor="password-was-login" className={styles.formInput}>
        Current password:
        <input
          id="password-was-login"
          placeholder="password"
          autoComplete="current-password"
          className={styles.input}
          value={currentPassword}
          onChange={(e) => setCurrentPassword(e.target.value)}
        />
      </label>
      <label htmlFor="password-login" className={styles.formInput}>
        New password:
        <input
          id="password-login"
          placeholder="new password"
          autoComplete="new-password"
          className={styles.input}
          value={newPassword}
          onChange={(e) => setNewPassword(e.target.value)}
        />
      </label>
      <button onClick={changePasswordHandle} type="button" className={styles.changeInfo}>
        Change
      </button>
    </div>
  );
}
