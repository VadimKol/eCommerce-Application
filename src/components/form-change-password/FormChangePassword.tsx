import type { ClientResponse, Customer, MyCustomerChangePassword } from '@commercetools/platform-sdk';
import { zodResolver } from '@hookform/resolvers/zod';
import { useForm } from 'react-hook-form';
import { toast } from 'react-toastify';

import { tokenCache } from '@/api/build-client';
import { changePassword, login } from '@/api/client-actions.ts';

import { Tooltip } from '../tooltip/Tooltip.tsx';
import { type FormValues, registerSchema } from './register-schema.ts';
import styles from './styles.module.scss';
import type { FormChangePasswordProps } from './types.ts';

export function FormChangePassword({ email, version, setPersonInfo }: FormChangePasswordProps): JSX.Element {
  const {
    register,
    getFieldState,
    watch,
    formState: { errors, isValid },
  } = useForm<FormValues>({ mode: 'onChange', resolver: zodResolver(registerSchema) });

  const {
    onChange: onChangeCurrentPassword,
    name: CurrentPassword,
    ref: refCurrentPassword,
  } = register('currentPassword');
  const { onChange: onChangeNewPassword, name: NewPassword, ref: refNewPassword } = register('newPassword');

  const currentPasswordValue = getFieldState('currentPassword');
  const newPasswordValue = getFieldState('newPassword');

  let currentPasswordClass = styles.input;
  let newPasswordClass = styles.input;

  if (currentPasswordValue.isDirty) {
    currentPasswordClass += currentPasswordValue.invalid ? ` ${styles.invalid}` : ` ${styles.valid}`;
  }

  if (newPasswordValue.isDirty) {
    newPasswordClass += newPasswordValue.invalid ? ` ${styles.invalid}` : ` ${styles.valid}`;
  }

  const currentPassword = watch('currentPassword');
  const newPassword = watch('newPassword');

  const changePasswordHandle = (): void => {
    const myCustomerChangePassword: MyCustomerChangePassword = {
      version,
      currentPassword,
      newPassword,
    };
    changePassword(myCustomerChangePassword)
      .then((response: ClientResponse<Customer>) => {
        if (response) {
          toast('Previous password entered correctly', { type: 'success' });
          setPersonInfo({
            version: response.body.version || 1,
            firstName: response.body.firstName || '',
            lastName: response.body.lastName || '',
            dateOfBirth: response.body.dateOfBirth || '',
            email: response.body.email || '',
            password: response.body.password || '',
            defaultShippingAddressId: response.body.defaultShippingAddressId || '',
            defaultBillingAddressId: response.body.defaultBillingAddressId || '',
            billingAddressIds: response.body.billingAddressIds || [],
            shippingAddressIds: response.body.shippingAddressIds || [],
          });
        }
      })
      .then(() => {
        const password = myCustomerChangePassword.newPassword;

        login({ email, password })
          .then(() => {
            localStorage.setItem('geek-shop-token', `${tokenCache.get().token}`);
            toast(`Password change was successful`, { type: 'success' });
          })
          .catch(() => {
            toast('An error occurred when changing your password, please try again', { type: 'error' });
          });
      })
      .catch((err: Error) => {
        toast(`Password error ${err.message}`, { type: 'error' });
      });
  };

  return (
    <form
      className={styles.form}
      onSubmit={(event) => {
        event.preventDefault();
        if (isValid) {
          changePasswordHandle();
        } else {
          toast('Validation error', { type: 'error' });
        }
      }}
    >
      <div className={styles.blockPassword}>
        <div className={styles.titleWithTooltip}>
          <h2>Change password</h2>
          <Tooltip
            text=" To change your password, enter your current password and the new one. Once the current password has been
        successfully verified, it will be replaced with a new password."
          />
        </div>
        <div className={styles.inputWithError}>
          <label htmlFor="password-was-login" className={styles.formInput}>
            Current password:
            <input
              onChange={(event) => {
                onChangeCurrentPassword(event).catch(() => {});
              }}
              id="current-password"
              placeholder="password"
              autoComplete="current-password"
              className={currentPasswordClass}
              ref={refCurrentPassword}
              aria-invalid={errors.currentPassword || !currentPasswordValue.isDirty ? 'true' : 'false'}
              name={CurrentPassword}
            />
          </label>
          {errors.currentPassword && (
            <span role="alert" className={styles.errorMsg}>
              {errors.currentPassword.message}
            </span>
          )}
        </div>
        <div className={styles.inputWithError}>
          <label htmlFor="password-login" className={styles.formInput}>
            New password:
            <input
              onChange={(event) => {
                onChangeNewPassword(event).catch(() => {});
              }}
              id="new-password"
              placeholder="new password"
              autoComplete="new-password"
              className={newPasswordClass}
              ref={refNewPassword}
              aria-invalid={errors.currentPassword || !newPasswordValue.isDirty ? 'true' : 'false'}
              name={NewPassword}
            />
          </label>
          {errors.newPassword && (
            <span role="alert" className={styles.errorMsg}>
              {errors.newPassword.message}
            </span>
          )}
        </div>
        <button type="submit" className={!isValid ? `${styles.changeInfo} ${styles.disabled}` : styles.changeInfo}>
          Change
        </button>
      </div>
    </form>
  );
}
