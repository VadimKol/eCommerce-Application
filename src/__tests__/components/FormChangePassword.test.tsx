import { render } from '@testing-library/react';

import { FormChangePassword } from '@/components/form-change-password/FormChangePassword';

it('renders correctly', () => {
  const setter = (): void => {};

  const { container } = render(<FormChangePassword email="user@example.com" version={1} setPersonInfo={setter} />);
  expect(container).toMatchSnapshot();
});
