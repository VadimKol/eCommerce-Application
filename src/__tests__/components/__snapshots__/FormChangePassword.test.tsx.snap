// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <form
    class="form"
  >
    <div
      class="blockPassword"
    >
      <div
        class="titleWithTooltip"
      >
        <h2>
          Change password
        </h2>
        <div
          class="tooltip"
        >
          <span
            class="tooltipText"
          >
             To change your password, enter your current password and the new one. Once the current password has been
        successfully verified, it will be replaced with a new password.
          </span>
        </div>
      </div>
      <div
        class="inputWithError"
      >
        <label
          class="formInput"
          for="password-was-login"
        >
          Current password:
          <input
            aria-invalid="true"
            autocomplete="current-password"
            class="input"
            id="current-password"
            name="currentPassword"
            placeholder="password"
          />
        </label>
      </div>
      <div
        class="inputWithError"
      >
        <label
          class="formInput"
          for="password-login"
        >
          New password:
          <input
            aria-invalid="true"
            autocomplete="new-password"
            class="input"
            id="new-password"
            name="newPassword"
            placeholder="new password"
          />
        </label>
      </div>
      <button
        class="changeInfo disabled"
        type="submit"
      >
        Change
      </button>
    </div>
  </form>
</div>
`;
